android/src/main/kotlin/com/signify/hue/flutterreactiveble/ble/BleWrapper.kt

```diff
@@ -3,7 +3,7 @@ package com.signify.hue.flutterreactiveble.ble
 import com.polidea.rxandroidble2.RxBleConnection
 import java.util.UUID
 
-data class ScanInfo(val deviceId: String, val name: String, val rssi: Int, val serviceData: Map<UUID, ByteArray>)
+data class ScanInfo(val deviceId: String, val name: String, val rssi: Int, val serviceData: Map<UUID, ByteArray>, val manufacturerData: Map<Int, ByteArray>)
```

Comment:
For the record, I'm not really happy about bloating even more this intermediate structure.
We may want to consider just passing around the raw advertised bytes and have an "advertisement parser" on the Dart side.

Reply:
Yess indeed we should just parse the raw advertisement data. Let's create an issue for it

Reply:
Which issue? Let's not integrate what creates an issue.

I checked the Manufacturer Specific Data format and I do not understand, how a map is a suitable representation for it. According to the spec it's just a uint16 manufacturer identifier and arbitrary binary payload following. Let's stick to this exact representation.

Reply:
Technically there could have multiple manufacturer specific data sequences in a single advertisement packet.
The iOS API only expose one (and the API doesn't split automatically the identifier code with the payload).

But the Android API (and the Web Bluetooth API) exposes this as a Map<int, bytes>.

I can switch to the iOS model and just do the concatenation on Android.

My comment here was more about the way we are passing data around in Android side with an useless temporary conversion.

Reply:
I didn't realize that Android uses `Map<int, bytes>` representation. We can use it, as it's more capable and accommodates the iOS data as well. Is a _map_ here necessary, though? What is the meaning of a key in that map? (if it's just an index, we can use a list; I don't think we should concatenate the data)

Reply:
From the Android docs you @xvrh have referred to it looks like the actual type for the aggregate manufacturer specific data is `SparseArray<byte[]>`. Could you clarify why do you suggest using a map, instead of a list for it?

Reply:
The key is the Company Identifier code (https://www.bluetooth.com/specifications/assigned-numbers/company-identifiers/)
This is the first 2 bytes of the manufacturer data sequence.

We can also a `List<Uint8List>` and the user will be responsible to split the company identifier and the payload.
On iOS it will have maximum one entry in the list.

On Android, we would have to loop over the entry in manufacturerSpecificData and recreate the raw payload:
```kotlin
for (i in 0..manufacturerSpecificData.size()) {
    val companyId = manufacturerSpecificData.keyAt(i)
    val payload = manufacturerSpecificData.get(companyId);

    val rawData = mutableListOf<Byte>()
    rawData.add((companyId.toByte()))
    rawData.add(((companyId.shr(8)).toByte()))
    rawData.addAll(2, payload.asList());

    datas.add(rawData);
}
```

I'm fine with all the options

Reply:
> The key is the Company Identifier code

Does the spec guarantee that multiple manufacturer specific data sections never share the same manufacturer identifier?

Reply:
@werediver despite its name, `SparseArray` is actually the equivalent of a `Map` with integer keys :-)
https://developer.android.com/reference/android/util/SparseArray

Reply:
> Does the spec guarantee that multiple manufacturer specific data sections never share the same manufacturer identifier?
I'm really not qualified to answer that. 
I guess any device can advertise whatever he wants. It just that the phone won't have the api to access the data correctly.

Reply:
> I guess any device can advertise whatever he wants. It just that the phone won't have the api to access the data correctly.

I'd reason the same way. But that means we cannot pack multiple manufacturer specific data sections into a map using their manufacturer identifiers as the map keys, right? If the current implementation is doing that, that's a problem.

Reply:
> If the current implementation is doing that, that's a problem.
The Android Bluetooth API is doing it.
But they also expose the raw bytes, so any one can still access all the data.

On iOS, only one record is exposed anyway.

Reply:
@xvrh What would you say to the following proposition? Let's indeed stick to the simplified approach, as in iOS. That must cover for 99% of the cases and we'll have a possibility to change / extend the implementation later, if there is demand.

Reply:
@werediver OK fine with me.
I'll push an update with the user-facing api be `Uint8List` containing a single manufacturer data entry with the identifier code as the first 2 bytes.